### ISSUE TITLE
[4.0][Processes] Implement subscriptions/Pre Orders handling

### ISSUE THREAD

**Issue by roykho** [woocommerce/woocommerce-gateway-stripe#290]
Currently Subscriptions and Pre Orders work with credit card payments. Now that we have introduced many different payment methods, we need to evaluate how to handle them.

Not all payment methods will support recurring charges. And it also depends on the bank the customer chooses to use. We can determine this only upon the charge during checkout to know if the customer provided payment method will support recurring charges or not.

So at this point we need to make a decision on how to handle this flow. A little background on recurring charges -- currently some of our other gateways also does not support recurring charges and merchants are relying on manual process. So when it comes time for payment, they send out payment links to customers to ask for a renewal. Both Subscriptions and Pre Orders support this manual style of recurring charges.

For Stripe, how do we want to handle this? Do we do it the same way as how it works with other gateways with no recurring charges support or do we stop the customer from purchasing and ask for a different form of payment? Note that some customers only have one form of payment depending on their location.

I am personally in favor of doing it how we're currently doing it with other gateways and prevent the checkout flow from stopping. Process the order first. If so, then how do we communicate to the merchant and customer that their payment for a recurring charge product will need to be manually paid again when time comes.

cc @mattyza

**Comment by roykho** [comment]
So as it turns out we may need to hold off on this implementation. Of all the payment methods added, only two of them (SEPA and Alipay) "may" potentially be able to recur. Reference here https://stripe.com/payments/payment-methods-guide

Upon speaking with Thorsten, Alipay is still in beta for its recurring charge and for SEPA, there are requirements which makes it impossible for us to implement. The requirements for SEPA is as follows:

> SEPA Direct Debit is reusable by default and can be activated on accounts with active payment history: "Before you can accept SEPA Direct Debit payments, your Stripe account needs to have regularly processed card payments for 60-75 days. If you would like to discuss starting sooner, please get in touch. Refer to our guide on SEPA Direct Debit payments guide to learn more about its requirements." https://stripe.com/docs/sources/sepa-debit

So there is no way for our extension to know that kind of information. So I propose just allow manual recurring charges for now on these added payment methods. We can revisit this again when something changes on Stripe's end.

**Comment by roykho** [comment]
Update: So Thorsten stated for SEPA, it will just return not reusable if it doesn't meet the criteria above so we can still implement it as long as there is logic to handle when it is NOT reusable. For those cases, we can just process the one time payment as normal and have subs manual recur the payment.

**Comment by itsmrwalsh** [comment]
Does any of this discussion relate to the core Stripe payment methods (Elements/Checkout/JS) Or will those still be able to utilize recurring payments?

**Comment by roykho** [comment]
@itsmrwalsh - does not affect the core payment method of credit cards.

**Comment by unixtam** [comment]
@roykho, the SEPA Direct Debit requirements need to be met before any SEPA Direct Debit payments can be made at all, whether subscription or single payment. The admin needs to request activation of SEPA via the Stripe dashboard. Once activated, it appears to support recurring payments out of the box.

So unless I'm missing something, it would seem that if this plugin supports single SEPA Direct Debit payments, SEPA Direct Debit subscriptions should work just fine.

See https://stripe.com/docs/sources/sepa-debit

**Comment by roykho** [comment]
@unixtam SEPA is already compatible with our Subscriptions.

**Comment by unixtam** [comment]
@roykho, thanks for clearing that up. I found this issue as the last open item in the "Elements & Sources Support" project backlog and thought I'd comment to help move things along.

**Comment by roykho** [comment]
@unixtam I believe you were confusing Stripe's subscriptions (recurring payments) with WooCommerce Subscriptions extension. They're not the same. This opened issue was to address all payments that didn't yet support that subscription extension. SEPA DD now does support it, but others still do not.

**Comment by unixtam** [comment]
@roykho I did assume this issue was about WCS compatibility. The confusion was that SEPA DD for Subscriptions seemed to be problematic, which it shouldn't be, given that SEPA DD for single payments were already implemented as per https://github.com/woocommerce/woocommerce-gateway-stripe/issues/296. The only mention of SEPA DD Subscriptions I could find was closed  *"Not implemented yet."* (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/322). Combing through the 4.0 branch commits, however, I now found https://github.com/woocommerce/woocommerce-gateway-stripe/commit/479fad36ed60e69ec42ca10acfd4c5f628d8c68a. Looking good.
